proposals/NNNN-Keyboard-Enhancements.md

```diff
@@ -0,0 +1,219 @@
+        <param name="numOfKeys" type="Integer" mandatory="true"/>
+    </struct>
+```
+
+
+
+#### Change in _"KeyboardLayout"_ enum
+
+```xml
+<enum name="KeyboardLayout">
+	:
+	<element name="NUMERIC"/>
+</enum>
+```
+
+#### Change in _"KeyboardProperties"_ struct
+
+1. App can request masking input characters using _"maskInputCharacters"_ param in _"KeyboardProperties"_ struct.
+2. App can request changes in special characters shown in the keyboard layout using _"customizeKeys"_ array. To set these buttons, app can just send an array of _"Strings"_ with special characters in _"customizeKeys"_ in _"SetGlobalProperties"_ RPC.
+3. If App does not send this string, HMI should show default special characters in the keyboard layout.
+4. If the number of keys in _"customizeKeys"_ array is more than customizable keys allowed, the SDL core should respond with _"INVALID_DATA"_ and the info string should include a detailed message, that **_"customizeKeys exceeds the number of customizable keys in this Layout"_**.
+5. If the number of keys in _"customizeKeys"_ array is less than or equal to customizable keys allowed, the SDL core should respond with _"SUCCESS"_. 
+6. HMI should show default characters in the remaining customizable Keys. if _"customizeKeys"_ array is less than or equal to the customizable keys allowed.
+7. If a certain special character is not supported by the system. The HMI should send _"WARNING"_ response, with _"info"_ text as _" some symbols might not be supported by system"_. This will give a chance to use symbols that are supported and also inform the app about the system not supporting certain characters.
+
+**Example usage:** If an app wants to change three symbols, the app sends the value of _"customizeKeys"_ as _"₹£$"_. The parameter follows a similar structure as _"limitedCharacterList"_.
+ 
+ ```xml
+     <struct name="KeyboardProperties" since="3.0">
+        <description>Configuration of on-screen keyboard (if available).</description>         
+		:       
+		<param name="maskInputCharacters" type="KeyboardInputMask" mandatory="false"  since="X.X"> 
+			<description>Allows an app to mask entered chanracters on HMI</description>
+		</param> 
+		<param name="customizeKeys" type="String" maxlength="1" minsize="1" maxsize="10" array="true" mandatory="false"  since="X.X">
+            <description>Array of special characters to show in customisable Keys.</description>
+            <description>If omitted, keyboard will show default special characters</description>
+        </param>  
+    </struct>
+ ```
+
+#### Change in _"RegisterAppInterface"_ response
+ 
+```xml
+<function name="RegisterAppInterface" functionID="RegisterAppInterfaceID" messagetype="response" since="1.0">
+		:
+        <param name="keyboardCapabilities"  type="KeyboardCapabilities" mandatory="false" since="X.X">
+            <description>See KeyboardCapabilities</description>
+        </param>
+</function>
+```
+

```

Comment:
Please add the `## Potential downsides` and `## Impact on existing code` sections to this proposal.  See [template](https://github.com/smartdevicelink/sdl_evolution/blob/master/0000-template.md) for more information.

Reply:
Added changes in c845c7115e0d5fa37dbe6ca15e12aceddf02db16